src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs

```diff
@@ -127,13 +129,40 @@ public IHost Build()
             }
             _hostBuilt = true;
 
+            // REVIEW: If we want to raise more events outside of these calls then we will need to
+            // stash this in a field.
+            using var diagnosticListener = new DiagnosticListener("Microsoft.Extensions.Hosting");
+            const string hostBuildingEventName = "HostBuilding";
+            const string hostBuiltEventName = "HostBuilt";
+
+            if (diagnosticListener.IsEnabled() && diagnosticListener.IsEnabled(hostBuildingEventName))
+            {
+                Write(diagnosticListener, hostBuildingEventName, this);
+            }
+
             BuildHostConfiguration();
             CreateHostingEnvironment();
             CreateHostBuilderContext();
             BuildAppConfiguration();
             CreateServiceProvider();
 
-            return _appServices.GetRequiredService<IHost>();
+            var host = _appServices.GetRequiredService<IHost>();
+            if (diagnosticListener.IsEnabled() && diagnosticListener.IsEnabled(hostBuiltEventName))
+            {
+                Write(diagnosticListener, hostBuiltEventName, host);
+            }
+
+            return host;
+        }
+
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:UnrecognizedReflectionPattern",
+            Justification = "The values being passed into Write are being consumed by the application already.")]
```

Comment:
It isn't the "The values being passed into Write" that are the issue, but instead the recursive properties of the values.

DiagnosticSource has an EventSource bridge that allows end-users to set up "filter specs" that specify which properties of the objects passed here should be written to the EventSource. If those properties aren't used in the app, they could be trimmed.

Reply:
I think that's true in general but in the use case for these events, I'm not worried:
- There won't be any trimming happening in development. You don't publish an application trimmed and then run tools in it to grab the IHost and IHostBuilder.
- In the 90% case, the relevant properties/methods are going to be used as part of application bootstrapping.

Reply:
Of course, if we don't expect people to be using the EventSource bridge, that could be a "good enough" reason to suppress.

Reply:
> There won't be any trimming happening in development. You don't publish an application trimmed and then run tools in it to grab the IHost and IHostBuilder.

What about someone else who takes a dependency on this new DiagnosticSource?

Reply:
> Of course, if we don't expect people to be using the EventSource bridge, that could be a "good enough" reason to suppress.

Yes this isn't a use case. There's nothing serializable on there that can be sent out of process.

> What about someone else who takes a dependency on this new DiagnosticSource?

Then we'll learn about the scenario then because it won't work for that scenario we don't understand.

Reply:
I think anyone else who used the DiagnosticSource would either:
a) cast the received object to its correct type which would make the linker aware of the usage - everything works nicely here
b) private reflection - I trust this puts the onus on the new DiagnosticSource consumer to ensure the fields aren't removed